package/MDAnalysis/analysis/MSD.py

```diff
@@ -0,0 +1,398 @@
+                 dt_restart, out_name='msd', len_msd=250, write_output=True,
+                 max_data=False):
+        """
+        Parameters
+        ----------
+        universe : object
+            Universe object containing trajectory data
+        select_list : list
+            List of selection strings
+        start_frame : int
+            Time when analysis begins (in frames)
+        stop_frame : int
+            Time when analysis ends (in frames)
+        dt_restart : int
+            Spacing of restarts (in frames)
+        out_name: str
+            Name of pickle output file: (out_name).p
+        len_msd : int
+            Specifies the maximum tau to be calculated (frames)
+            (saves memory in array allocation)
+        write_output : bool
+            Specifies whether to write pickle file
+        max_data : bool
+            If true, include tau values after stop_frame in calculation (this
+            is useful for avoiding data loss when breaking large calculations
+            across multiple batch jobs)
+        """
+
+        self.universe = universe
+        self.select_list = select_list
+        if isinstance(select_list, str):
+            self.select_list = select_list.split(',')
+        print(self.select_list)
+        self.start_frame = int(start_frame)
+        self.stop_frame = int(stop_frame)
+        self.dt_restart = int(dt_restart)
+        self.out_name = out_name
+        # handling short simulation cases
+        n_frames = int(stop_frame) - int(start_frame) + 1
+        if int(len_msd) > int(n_frames):
+            len_msd = int(n_frames)
+        self.len_msd = int(len_msd)
+        self.write_output = write_output
+        self.max_data = max_data
+
+        # initializing object attributes
+        self.pos, self.atomids_per_sel_at_t, self.map_atomids_to_pos_at_t, = \
+            self._init_pos_deque()
+
+        n_sel = len(self.pos)
+        self.msd = np.zeros((n_sel, self.len_msd), dtype='float64')
+        self.n_samples = np.zeros((n_sel, self.len_msd), dtype='int')
+
+        # check that trajectory length is correct
+        try:
+            assert len(self.universe.trajectory) >= \
+                    (self.stop_frame-self.start_frame+1)
+        except RuntimeError:
+            print("Sample interval exceeds trajectory length. This may result"
+                  "from choice of start_frame/stop_frame or insufficient RAM"
+                  "when loading the trajectory.")
+
+    def _select(self):
+        """Generates list of atom groups that satisfy select_list strings
+
+        Returns
+        -------
+        selections : list
+            list of atom groups
+        """
+        selections = []
+        for i, sel in enumerate(self.select_list):
+            selections.append(self.universe.select_atoms(sel))
+            if selections[i].n_atoms == 0:
+                selections.pop(i)
+                selections.append(None)
+        
```

Comment:
This can be optimized even further. Because python uses a pbc condition for array indices you can use `[-1]` to access the last element. Also the pop/append is unnecessary if we just set the value directly.

```python
for sel in self.select_list:
    selections.append(self.universe.select_atoms(sel))
    if selections[-1].n_atoms == 0:
        selections[-1] = None
```